File: day3.pl
#!/usr/bin/perl

use strict;
use warnings;

sub tonum
{
  my $val = shift;
  my $result = 0;
  my $i = 0;
  while( $i < length($val))
  {
    $result *= 2;
    if(substr($val, $i++, 1) == '1')
    {
      $result++;
    }
  }
  return $result;
}

sub most
{
  my $ind = shift;
  my @input = @_;
  my $i = 0;
  my $ones = 0;

  while($i < scalar(@input))
  {
    if(substr($input[$i++], $ind, 1) == '1')
    {
      $ones++;
    }
  }
  if(($ones * 2) >= scalar(@input))
  {
    return 1;
  }
  return 0;
}

sub least
{
  my $ind = shift;
  my @input = @_;
  my $i = 0;
  my $ones = 0;
  
  while($i < scalar(@input))
  {
    if(substr($input[$i++], $ind, 1) == '1')
    {
      $ones++;
    }
  }
  if(($ones * 2) < scalar(@input))
  {
    return 1;
  }
  return 0;
}

sub filter
{
  my $bit = shift;
  my @input = @_;
  my $value = '';
  my $i = 0;

  while($i < length($input[0]))
  {
    my $count = scalar(@input);
    my $cbit;
    my $j = 0;
    my $count2 = 0;
    my @input2 = ();

    if($bit == 1)
    {
      $cbit = most($i, @input);
    } else {
      $cbit = least($i, @input);
    }
    $value .= ($cbit ? '1': '0');
    $i++;

    while($j < $count)
    {
      if(substr($input[$j], 0, $i) == $value)
      {
        $input2[$count2++] = $input[$j];
      }
      $j++;
    }
    @input = @input2;
    $count = $count2;
    if(scalar(@input) == 1)
    {
      return $input[0];
    }
  }
  return $value;
}

my @numbers = ();
my $count = 0;
my $gamma = 0;
my $epsilon = 0;
my $i = 0;
open FIN, '<'.$ARGV[0];
while(<FIN>)
{
  chomp;
  $numbers[$count++] = $_;
}
while($i < length($numbers[0]))
{
  $gamma *= 2;
  $epsilon *= 2;
  $gamma += most($i, @numbers);
  $epsilon += least($i, @numbers);
  $i++;
}
print "Part 1:", $gamma * $epsilon, "\n";
my $a = tonum(filter(1, @numbers));
my $b = tonum(filter(0, @numbers));
print "Part 2:", $a * $b, "\n";

File: day3.rb
def most(nums, index)
  ones = 0
  for i in 0..(nums.length - 1)
    if nums[i][index] == '1'
      ones += 1
    end
  end
  if (ones * 2) >= nums.length
    return 1
  end
  return 0
end

def least(nums, index)
  ones = 0
  for i in 0..(nums.length - 1)
    if nums[i][index] == '1'
      ones += 1
    end
  end
  if (ones * 2) < nums.length
    return 1
  end
  return 0
end

def filter(bit, nums)
  value = ''
  for i in 0..((nums[0].length) -1)
    count2 = 0
    nums2 = []
    if bit == 1
      cbit = most(nums, i)
    else
      cbit = least(nums, i)
    end
    value += '01'[cbit]
    for j in 0..(nums.length - 1)
      if nums[j][0..i] == value
        nums2[count2] = nums[j]
        count2 += 1
      end
    end
    nums = nums2
    if nums.length == 1
      return nums[0]
    end
  end
  return value
end

nums = []

File.open(ARGV[0]).each do |line|
  nums.append(line.strip)
end

gamma = 0
epsilon = 0
a = filter(1, nums)
b = filter(0, nums)
c = 0
d = 0
for i in 0..((nums[0]).length - 1)
  gamma *= 2
  epsilon *= 2
  c *= 2
  d *= 2
  gamma += most(nums, i)
  epsilon += least(nums, i)
  if a[i] == '1'
    c += 1
  end
  if b[i] == '1'
    d += 1
  end
end
puts "Part 1: #{gamma * epsilon}"
puts "Part 2: #{c * d}"

